src/main/java/org/jabref/logic/search/indexing/PostgreIndexer.java

```diff
@@ -0,0 +1,209 @@
+
+    public PostgreIndexer(BibDatabaseContext databaseContext, Connection connection) {
+        this.databaseContext = databaseContext;
+        this.connection = connection;
+        this.libraryName = databaseContext.getDatabasePath().map(path -> path.getFileName().toString()).orElse("unsaved");
+        if ("unsaved".equals(databaseContext.getPostgreTableName())) {
+            this.tableName = "unsaved" + NUMBER_OF_UNSAVED_LIBRARIES++;
+        } else {
+            this.tableName = databaseContext.getPostgreTableName();
+        }
+        setup();
+    }
+
+    /**
+     * Creates a table for the library in the database, and sets up indexes on the columns.
+     */
+    private void setup() {
+        try {
+            connection.createStatement().executeUpdate("""
+                    CREATE TABLE IF NOT EXISTS "%s" (
+                        %s TEXT NOT NULL,
+                        %s TEXT NOT NULL,
+                        %s TEXT,
+                        PRIMARY KEY (%s, %s)
+                    )
+                    """.formatted(tableName,
+                    PostgreConstants.ENTRY_ID,
+                    PostgreConstants.FIELD_NAME,
+                    PostgreConstants.FIELD_VALUE,
+                    PostgreConstants.ENTRY_ID,
+                    PostgreConstants.FIELD_NAME));
+            LOGGER.debug("Created table for library: {}", libraryName);
+
+            // btree index on id column
+            connection.createStatement().executeUpdate("""
+                    CREATE INDEX "%s" ON "%s" ("%s")
+                    """.formatted(PostgreConstants.ENTRY_ID.getIndexName(tableName), tableName, PostgreConstants.ENTRY_ID));
+
+            // btree index on field name column
+            connection.createStatement().executeUpdate("""
+                    CREATE INDEX "%s" ON "%s" ("%s")
+                    """.formatted(PostgreConstants.FIELD_NAME.getIndexName(tableName), tableName, PostgreConstants.FIELD_NAME));
+
+            // trigram index on field value column
+            connection.createStatement().executeUpdate("""
+                    CREATE INDEX "%s" ON "%s" USING gin ("%s" gin_trgm_ops)
+                    """.formatted(PostgreConstants.FIELD_VALUE.getIndexName(tableName), tableName, PostgreConstants.FIELD_VALUE));
+
+            LOGGER.debug("Created indexes for library: {}", libraryName);
+        } catch (SQLException e) {
+            LOGGER.error("Could not create table for library: {}", libraryName, e);
+        }
+    }
+
+    public void updateOnStart(BackgroundTask<?> task) {
+        addToIndex(databaseContext.getDatabase().getEntries(), task);
+    }
+
+    public void addToIndex(Collection<BibEntry> entries, BackgroundTask<?> task) {
+        task.setTitle(Localization.lang("Indexing bib fields for %0", libraryName));
+        int i = 1;
+        long startTime = System.currentTimeMillis();
+        LOGGER.debug("Adding {} entries to index", entries.size());
```

Comment:
When I was working on my AI features, we used file(s) for each mention of linked files.

Maybe in your case it can be entry(ies)?

Reply:
> Maybe in your case it can be entry(ies)?

This gets a little inconvenient to read, but nitpick - it's debug log anyway so both work, if you want to go with strict uniformity.

Reply:
> This gets a little inconvenient to read, but nitpick

We decided for that to enable more easy covering of Arabic... Full Arabic is hard. See https://phrase.com/blog/posts/pluralization/

---


    zero → “لم نزرع أي شجرة حتى الآن”
    one → “لقد زرعنا شجرة ١ حتى الآن”
    two → “لقد زرعنا شجرتين ٢ حتى الآن”
    few → “لقد زرعنا ٣ شجرات حتى الآن”
    many → “لقد زرعنا ١١ شجرة حتى الآن”
    other → “لقد زرعنا ١٠٠ شجرة حتى الآن”

---